File: scala/src/test/scala/sodium/MemoryTest4.scala
package sodium

object MemoryTest4 {
  def main(args: Array[String]): Unit = {
    new Thread() {
      override def run(): Unit = {
        try {
          while (true) {
            println(s"memory ${Runtime.getRuntime().totalMemory()}")
            Thread.sleep(5000)
          }
        } catch {
          case e: InterruptedException => e.printStackTrace()
        }
      }
    }.start()

    val et = new Stream[Int]()
    val eChange = new StreamSink[Int]()
    val oout = eChange.map(_ => et).hold(et)
    val out = Cell.switchS[Int](oout)
    val l = out.listen(tt => {
      println(tt)
    })
    var i = 0
    while (i < 1000000000) {
      eChange.send(i)
      i += 1
    }
    l.unlisten()
  }
}

File: java/src/test/java/nz/sodium/MemoryTest4.java
package nz.sodium;

public class MemoryTest4
{
    public static void main(String[] args)
    {
        new Thread() {
            public void run()
            {
                try {
                    while (true) {
                        System.out.println("memory "+Runtime.getRuntime().totalMemory());
                        Thread.sleep(5000);
                    }
                }
                catch (InterruptedException e) {
                    System.out.println(e.toString());
                }
            }
        }.start();

        StreamSink<Integer> et = new StreamSink<Integer>();
        StreamSink<Integer> eChange = new StreamSink<Integer>();
        Cell<Stream<Integer>> oout = eChange.map(x -> (Stream<Integer>)et).hold((Stream<Integer>)et);
        Stream<Integer> out = Cell.switchS(oout);
        Listener l = out.listen(tt -> {
            System.out.println(tt);
        });
        int i = 0;
        while (i < 1000000000) {
            eChange.send(i);
            i++;
        }
        l.unlisten();
    }
}
